File: frontend/src/api/importAccdb.ts
import { api } from './client'

const BASE = '/import'
// Для multipart/form-data upload api.* не підходить (він серіалізує JSON).
// Тому залишаємо raw fetch але вручну додаємо Authorization header.
const TOKEN_KEY = 'bakery_token'
function authHeaders(): Record<string, string> {
  const token = localStorage.getItem(TOKEN_KEY)
  return token ? { Authorization: `Bearer ${token}` } : {}
}

// ─── Types ────────────────────────────────────────────────────────────────────

export interface ColumnMap {
  access_col:   string
  target_field: string
  description:  string
}

export interface TableDetail {
  access_table: string | null
  target_table: string
  count:        number
  column_map:   ColumnMap[]
  sample:       Record<string, string | null>[]
  warnings:     string[]
}

export interface PriceCategory {
  access_id:    string
  name:         string
  price_count:  number
  client_count: number
}

export interface AccdbPreview {
  temp_file_token:       string
  access_tables:         string[]
  product_types:         string[]       // unique values of 'Тип' from _Вироби
  price_categories:      PriceCategory[]
  base_price_category:   string         // auto-detected base category id
  routes:    TableDetail
  clients:   TableDetail
  products:  TableDetail
  prices:    TableDetail
  orders:    TableDetail
  finances:  TableDetail
  all_routes:               RoutePreview[]
  all_clients_preview:      ClientPreview[]
  suggested_route_skips:    string[]    // route names to auto-skip
  suggested_non_customers:  SuggestedNonCustomer[]
}

export interface SuggestedNonCustomer {
  access_id:          number
  name:               string
  suggested_kind:     string
  suggested_merge_id: number | null
}

export interface RoutePreview {
  access_id: number
  name:      string
}

export interface ClientPreview {
  access_id: number
  name:      string
}

// ─── Mapping types ────────────────────────────────────────────────────────────

export interface RouteMapping {
  access_id:     number
  import_it:     boolean    // false = skip this route
  name_override: string
  sort_order:    number
}

export interface CategoryMapping {
  access_type:   string    // value of 'Тип' in Access ('Хліб', 'Булка', …)
  category_name: string
  is_baked:      number    // 1 = baked, 0 = shop only
  sort_order:    number
  reserve_pct:   number
}

export interface ClientMapping {
  access_id:   number
  client_kind: 'customer' | 'shop' | 'writeoff' | 'ration'
  merge_with:  number | null  // if set → use existing SQLite client_id
  skip:        boolean        // if true → don't create, exclude from all mappings
}

export interface ImportMapping {
  temp_file_token:      string
  db_password:          string
  transition_date:      string          // YYYY-MM-DD
  finance_months:       number
  order_days:           number
  route_mappings:       RouteMapping[]
  category_mappings:    CategoryMapping[]
  client_mappings:      ClientMapping[]
  default_client_kind:  string
  base_price_category:  string          // Access КодКатегорії for base prices
  invoice_draft_from:   string | null   // YYYY-MM-DD; накладні з цієї дати = draft
}

// ─── Context types ────────────────────────────────────────────────────────────

export interface ExistingClient {
  id:          number
  full_name:   string
  short_name:  string | null
  client_kind: string
}

export interface ExistingRoute {
  id:         number
  name:       string
  sort_order: number
}

export interface ExistingCategory {
  id:         number
  name:       string
  is_baked:   number
  sort_order: number
}

export interface ImportContext {
  existing_clients:    ExistingClient[]
  existing_routes:     ExistingRoute[]
  existing_categories: ExistingCategory[]
}

// ─── Report types ─────────────────────────────────────────────────────────────

export interface EntityReport {
  found:        number
  imported:     number
  skipped:      number
  skip_reasons: Record<string, number>
  warnings:     string[]
  errors:       string[]
  notes:        string
}

export interface BalanceMismatch {
  client_id:        number
  client_name:      string
  access_balance:   number
  computed_balance: number
  diff:             number
}

export interface ImportPriceRange {
  price:      number
  valid_from: string
  valid_to:   string | null
}

export interface ClientPriceGroup {
  client_id:   number
  client_name: string
  ranges:      ImportPriceRange[]
}

export interface ZeroPriceProduct {
  id:            number
  name:          string
  client_groups: ClientPriceGroup[]
}

export interface ValidationReport {
  balance_mismatches:  BalanceMismatch[]
  zero_price_products: ZeroPriceProduct[]
  order_count_ok:      boolean
  overall_ok:          boolean
}

export interface ImportReport {
  success:         boolean
  started_at:      string
  finished_at:     string
  transition_date: string
  entities:        Record<string, EntityReport>
  validation:      ValidationReport
}

export interface ImportStatus {
  running:  boolean
  step:     string
  progress: number
  error:    string | null
}

// ─── API calls ────────────────────────────────────────────────────────────────

export async function uploadAccdb(file: File, password = ''): Promise<AccdbPreview> {
  const form = new FormData()
  form.append('file', file)
  if (password) form.append('password', password)
  const res = await fetch(`/api/v1${BASE}/upload`, {
    method: 'POST',
    body: form,
    headers: authHeaders(),
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }))
    throw new Error(err.detail ?? 'Помилка завантаження файлу')
  }
  return res.json()
}

export async function getImportContext(): Promise<ImportContext> {
  return api.get<ImportContext>(`${BASE}/context`)
}

export async function runImport(mapping: ImportMapping): Promise<{ status: string }> {
  try {
    return await api.post<{ status: string }>(`${BASE}/run`, mapping)
  } catch (e) {
    // У api.* exception формат: "STATUS STATUSTEXT: BODY". Пробуємо витягти detail.
    const msg = e instanceof Error ? e.message : String(e)
    const match = msg.match(/:\s*(\{.*\})/)
    if (match) {
      try {
        const body = JSON.parse(match[1])
        const detail = body.detail
        const niceMsg = Array.isArray(detail)
          ? detail.map((d: { msg?: string }) => d.msg ?? JSON.stringify(d)).join('; ')
          : (detail ?? 'Помилка запуску імпорту')
        throw new Error(niceMsg)
      } catch { /* fallthrough */ }
    }
    throw e
  }
}

export async function getImportStatus(): Promise<ImportStatus> {
  return api.get<ImportStatus>(`${BASE}/status`)
}

export async function getImportResult(): Promise<ImportReport | null> {
  try {
    return await api.get<ImportReport>(`${BASE}/result`)
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e)
    if (msg.startsWith('404')) return null
    throw e
  }
}

File: backend/schemas/import_accdb.py
from __future__ import annotations
from typing import Any, Literal
from pydantic import BaseModel, Field


class ColumnMap(BaseModel):
    """Одна колонка: Access → нова система."""
    access_col:   str        # точна назва колонки в Access
    target_field: str        # назва поля в SQLite
    description:  str = ""  # людиночитаний опис українською


class TableDetail(BaseModel):
    """Деталі однієї сутності: Access таблиця → SQLite таблиця."""
    access_table: str | None = None   # точна назва таблиці в Access (None = не знайдено)
    target_table: str = ""            # цільова таблиця
    count:        int = 0             # кількість рядків в Access
    column_map:   list[ColumnMap] = []
    sample:       list[dict[str, Any]] = []  # до 3 зразкових рядків (Access значення)
    warnings:     list[str] = []


class PriceCategory(BaseModel):
    """Цінова категорія з Access (_Категорії)."""
    access_id:    str   # значення КодКатегорії ("1", "9", "10", …)
    name:         str   # назва категорії
    price_count:  int = 0
    client_count: int = 0


class RoutePreview(BaseModel):
    access_id: int
    name:      str


class ClientPreview(BaseModel):
    access_id: int
    name:      str


class AccdbPreview(BaseModel):
    temp_file_token: str
    access_tables:       list[str] = []
    product_types:       list[str] = []          # унікальні значення поля 'Тип' з _Вироби
    price_categories:    list[PriceCategory] = [] # цінові категорії з _Категорії
    base_price_category: str = ""                # auto-detected Access id базової категорії

    routes:    TableDetail = Field(default_factory=lambda: TableDetail(target_table="routes"))
    clients:   TableDetail = Field(default_factory=lambda: TableDetail(target_table="clients"))
    products:  TableDetail = Field(default_factory=lambda: TableDetail(target_table="products"))
    prices:    TableDetail = Field(default_factory=lambda: TableDetail(target_table="prices"))
    orders:    TableDetail = Field(default_factory=lambda: TableDetail(target_table="orders"))
    finances:  TableDetail = Field(default_factory=lambda: TableDetail(target_table="finances"))
    stock:     TableDetail = Field(default_factory=lambda: TableDetail(target_table="shop_reconciliation_lines"))

    # Повні списки для wizard-маппінгу
    all_routes:           list[RoutePreview]  = []  # всі маршрути Access
    all_clients_preview:  list[ClientPreview] = []  # всі клієнти Access (id + name)

    # Авто-пропозиції для маппінгу
    suggested_route_skips:   list[str]  = []   # назви маршрутів для auto-skip (system, ПЕКАРНЯ)
    suggested_non_customers: list[dict] = []   # [{access_id, name, suggested_kind, suggested_merge_id}]


# ─── Маппінги для нового wizard ───────────────────────────────────────────────

class RouteMapping(BaseModel):
    """Маппінг одного маршруту Access."""
    access_id:    int
    import_it:    bool = True      # False = пропустити (не імпортувати)
    name_override: str = ""        # порожньо = використати оригінальну назву
    sort_order:   int = 0


class CategoryMapping(BaseModel):
    """Маппінг типу виробу Access → категорія нової системи."""
    access_type:   str            # значення 'Тип' в Access ('Хліб', 'Булка', …)
    category_name: str            # назва категорії в новій системі
    is_baked:      int = 1        # 1 = печеться, 0 = лише магазин
    sort_order:    int = 0
    reserve_pct:   float = 5.0


class ClientMapping(BaseModel):
    """Маппінг одного клієнта Access."""
    access_id:   int
    client_kind: str = 'customer'  # customer|shop|writeoff|ration
    merge_with:  int | None = None # якщо задано — не створювати, використати існуючий SQLite client_id
    skip:        bool = False      # якщо True — не створювати і не включати в жоден маппінг


class ImportMapping(BaseModel):
    temp_file_token:  str
    db_password:      str = ''
    transition_date:  str                           # YYYY-MM-DD
    finance_months:   int = Field(0, ge=0)   # 0 = вся історія
    order_days:       int = Field(0, ge=0)   # 0 = вся історія
    route_mappings:   list[RouteMapping] = []
    category_mappings: list[CategoryMapping] = []  # замінює product_type_categories
    client_mappings:  list[ClientMapping] = []     # замінює client_kinds
    default_client_kind: str = 'customer'
    base_price_category: str = ''                  # Access КодКатегорії для базових цін
    invoice_draft_from: str | None = None          # YYYY-MM-DD; накладні з цієї дати = draft


# ─── Звіт ─────────────────────────────────────────────────────────────────────

class EntityReport(BaseModel):
    found:        int = 0
    imported:     int = 0
    skipped:      int = 0
    skip_reasons: dict[str, int] = {}   # причина → кількість
    warnings:     list[str] = []
    errors:       list[str] = []
    notes:        str = ""


class BalanceMismatch(BaseModel):
    client_id:        int
    client_name:      str
    access_balance:   float
    computed_balance: float
    diff:             float


class ZeroPriceProduct(BaseModel):
    id:   int
    name: str


class ValidationReport(BaseModel):
    balance_mismatches:   list[BalanceMismatch] = []
    zero_price_products:  list[ZeroPriceProduct] = []
    order_count_ok:       bool = True
    overall_ok:           bool = True


class ImportReport(BaseModel):
    success:         bool
    started_at:      str
    finished_at:     str
    transition_date: str
    entities:        dict[str, EntityReport] = {}
    validation:      ValidationReport = ValidationReport()


# ─── Контекст для merge-маппінгу ──────────────────────────────────────────────

class ExistingClient(BaseModel):
    id:          int
    full_name:   str
    short_name:  str | None = None
    client_kind: str


class ExistingRoute(BaseModel):
    id:         int
    name:       str
    sort_order: int = 0


class ExistingCategory(BaseModel):
    id:         int
    name:       str
    is_baked:   int = 1
    sort_order: int = 0


class ImportContext(BaseModel):
    """Існуючі сутності БД для merge-маппінгу в wizard."""
    existing_clients:    list[ExistingClient] = []
    existing_routes:     list[ExistingRoute] = []
    existing_categories: list[ExistingCategory] = []
